core/src/main/java/org/infinispan/expiration/impl/ExpirationManagerImpl.java

```diff
@@ -107,6 +118,45 @@ public boolean isEnabled() {
       return enabled;
    }
 
+   @Override
+   public void handleInMemoryExpiration(InternalCacheEntry<K, V> entry) {
+      dataContainer.compute(entry.getKey(), ((k, oldEntry, factory) -> {
+         if (entry == oldEntry) {
+            // We have to delete from shared stores as well to make sure there are not multiple expiration events
+            persistenceManager.deleteFromAllStores(k, PersistenceManager.AccessMode.BOTH);
+            if (cacheNotifier != null) {
+               cacheNotifier.notifyCacheEntryExpired(k, entry.getValue(), entry.getMetadata());
+            }
+            return null;
+         }
+         return oldEntry;
+      }));
+   }
+
+   @Override
+   public void handleInStoreExpiration(K key) {
+      dataContainer.compute(key, ((k, oldEntry, factory) -> {
+         if (oldEntry == null || (oldEntry.canExpire() && oldEntry.isExpired(timeService.time()))) {
```

Comment:
Couldn't the entry be `null` because of another reason, e.g. it was removed explicitly?